lib/remap-parser/src/parser.lalrpop

```diff
@@ -0,0 +1,501 @@
+        "/" => Token::Operator("/"),
+        "%" => Token::Operator("%"),
+        "<" => Token::Operator("<"),
+        "<=" => Token::Operator("<="),
+        ">" => Token::Operator(">"),
+        ">=" => Token::Operator(">="),
+        "==" => Token::Operator("=="),
+        "!=" => Token::Operator("!="),
+
+        "&&" => Token::Operator("&&"),
+        "||" => Token::Operator("||"),
+        "??" => Token::Operator("??"),
+
+        "[" => Token::LBracket,
+        "{" => Token::LBrace,
+        "(" => Token::LParen,
+
+        "]" => Token::RBracket,
+        "}" => Token::RBrace,
+        ")" => Token::RParen,
+
+        "_r" => Token::InternalTest("_r"),
+        "_e" => Token::InternalTest("_e"),
+        "_m" => Token::InternalTest("_m"),
+        "_a" => Token::InternalTest("_a"),
+        "_q" => Token::InternalTest("_q"),
+        "_c" => Token::InternalTest("_c"),
+        "_as" => Token::InternalTest("_as"),
+        "_fn" => Token::InternalTest("_fn"),
+    }
+}
+
+// -----------------------------------------------------------------------------
+// parser entry points
+// -----------------------------------------------------------------------------
+
+// The main entrypoint into a VRL program.
+//
+// A program consists of one or more expressions.
+pub Program: Program = NonterminalNewline* <RootExprs> => Program(<>);
+
+// This nonterminal exists to aid in unit-testing. It exposes individual rules
+// through the "t ..." declaration, to allow testing individual rules without
+// having to generate parser functions for each rule, which kills build-times
+// and exposes entries into the parser that we don't want or need.
+pub Test: Test = {
+    // root
+    "_r" <Expr> => Test::Expr(<>),
+
+    // expressions
+    "_e" <Literal> => Test::Literal(<>),
+    "_e" <Container> => Test::Container(<>),
+
+    // arithmetic (math)
+    "_m" <ArithmeticExpr> => Test::Arithmetic(<>),
+
+    // atoms
+    "_a" <String> => Test::String(<>),
+    "_a" <Integer> => Test::Integer(<>),
+    "_a" <Float> => Test::Float(<>),
+    "_a" <Boolean> => Test::Boolean(<>),
+    "_a" <Null> => Test::Null(()),
+    "_a" <Regex> => Test::Regex(<>),
+
+    // collections
+    "_c" <Block> => Test::Block(<>),
+    "_c" <Array> => Test::Array(<>),
+    "_c" <Object> => Test::Object(<>),
+
+    // other
+    "_as" <Assignment> => Test::Assignment(<>),
+    "_fn" <FunctionCall> => Test::FunctionCall(<>),
+    "_q" <Query> => Test::Query(<>),
+};
+
+// -----------------------------------------------------------------------------
+// root expressions
+// -----------------------------------------------------------------------------
+
+RootExprs: Vec<Node<RootExpr>> = {
+    RootExpr => vec![<>],
+    (<RootExpr> EndOfExpression)*,
+};
+
+RootExpr: Node<RootExpr> = {
+    Expr => Node::new(<>.span(), RootExpr::Expr(<>)),
+
+    // Root expressions are allowed to fail. The parser will continue with the
+    // next expression in the program.
+    Sp<!> => {
```

Comment:
`!` is a special terminal that captures any error generated by the parser. This is what allows the compiler to continue compiling root expressions even if one of them fails.